// @cloudscape-design/code-view/test-utils/dom/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeViewWrapper = exports.ElementWrapper = void 0;
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
const dom_1 = require("@cloudscape-design/test-utils-core/dom");
Object.defineProperty(exports, "ElementWrapper", { enumerable: true, get: function () { return dom_1.ElementWrapper; } });
const utils_1 = require("@cloudscape-design/test-utils-core/utils");
const code_view_1 = require("./code-view");
exports.CodeViewWrapper = code_view_1.default;
dom_1.ElementWrapper.prototype.findCodeView = function (selector) {
    let rootSelector = `.${code_view_1.default.rootSelector}`;
    if ("legacyRootSelector" in code_view_1.default && code_view_1.default.legacyRootSelector) {
        rootSelector = `:is(.${code_view_1.default.rootSelector}, .${code_view_1.default.legacyRootSelector})`;
    }
    // casting to 'any' is needed to avoid this issue with generics
    // https://github.com/microsoft/TypeScript/issues/29132
    return this.findComponent(selector ? (0, utils_1.appendSelector)(selector, rootSelector) : rootSelector, code_view_1.default);
};
dom_1.ElementWrapper.prototype.findAllCodeViews = function (selector) {
    return this.findAllComponents(code_view_1.default, selector);
};
dom_1.ElementWrapper.prototype.findClosestCodeView = function () {
    // casting to 'any' is needed to avoid this issue with generics
    // https://github.com/microsoft/TypeScript/issues/29132
    return this.findClosestComponent(code_view_1.default);
};
function wrapper(root = document.body) {
    if (document && document.body && !document.body.contains(root)) {
        console.warn('[AwsUi] [test-utils] provided element is not part of the document body, interactions may work incorrectly');
    }
    ;
    return new dom_1.ElementWrapper(root);
}
exports.default = wrapper;
