// @cloudscape-design/components/test-utils/selectors/collection-preferences/content-display-preference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentDisplayOptionWrapper = void 0;
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
const selectors_1 = require("@cloudscape-design/test-utils-core/selectors");
const list_1 = require("../list");
const text_filter_1 = require("../text-filter");
const toggle_1 = require("../toggle");
const styles_selectors_js_1 = require("../../../collection-preferences/styles.selectors.js");
const getClassName = (suffix) => styles_selectors_js_1.default[`content-display-${suffix}`];
class ContentDisplayOptionWrapper extends selectors_1.ComponentWrapper {
    getListItem() {
        return new list_1.ListItemWrapper(this.getElement());
    }
    /**
     * Returns the drag handle for the option item.
     */
    findDragHandle() {
        return this.getListItem().findDragHandle();
    }
    /**
     * Returns the text label displayed in the option item.
     */
    findLabel() {
        return this.getListItem().findContent().findByClassName(styles_selectors_js_1.default['content-display-option-label']);
    }
    /**
     * Returns the visibility toggle for the option item.
     * Note that, despite its typings, this may return null for group items since groups do not have a visibility toggle.
     */
    findVisibilityToggle() {
        return this.getListItem().findContent().findComponent(`.${styles_selectors_js_1.default['content-display-option-toggle']}`, toggle_1.default);
    }
    /**
     * Returns all child option items nested under this item when it is a group.
     * Returns `null` when this item is a leaf column (has no nested children).
     *
     * The children are the leaf-level `ContentDisplayOptionWrapper`s inside the group's
     * nested `InternalList` — i.e. they already carry a drag handle and visibility toggle.
     *
     * @param option.group When `true`, returns only group items. When `false`, returns only leaf column items.
     *   When omitted, returns all child items regardless of type.
     */
    /* istanbul ignore next: :has() selector not supported in JSDOM */
    findChildrenOptions(option = {}) {
        const groupWrapper = this.getListItem().findContent().find('[data-item-type="group"]');
        if (!groupWrapper) {
            return null;
        }
        const nestedList = groupWrapper.find(`.${list_1.default.rootSelector}`);
        if (!nestedList) {
            return null;
        }
        const list = new list_1.default(nestedList.getElement());
        if (option.group === true) {
            return list.findAll(`li:has([data-item-type="group"])`).map(item => new ContentDisplayOptionWrapper(item.getElement()));
        }
        if (option.group === false) {
            return list.findAll(`li:has([data-item-type="column"])`).map(item => new ContentDisplayOptionWrapper(item.getElement()));
        }
        return list.findItems().map(item => new ContentDisplayOptionWrapper(item.getElement()));
    }
}
exports.ContentDisplayOptionWrapper = ContentDisplayOptionWrapper;
class ContentDisplayPreferenceWrapper extends selectors_1.ComponentWrapper {
    /**
     * Returns the title.
     */
    findTitle() {
        return this.findByClassName(getClassName('title'));
    }
    /**
     * Returns the preference description displayed below the title.
     */
    findDescription() {
        return this.findByClassName(getClassName('description'));
    }
    getList() {
        return new list_1.default(this.getElement());
    }
    /**
     * Returns an option for a given index.
     *
     * @param index 1-based index of the option to return.
     */
    findOptionByIndex(index) {
        const item = this.getList().findItemByIndex(index);
        return item && new ContentDisplayOptionWrapper(item.getElement());
    }
    /**
     * Returns the top-level items in the preference list.
     *
     * For tables **without** column grouping this returns all column options.
     * For tables **with** column grouping this returns the top-level entries only
     * (which are group items). Use `.findChildrenOptions()` on a group item to
     * access the leaf columns nested within it.
     *
     * @param option.group When `true`, returns only group items. When `false`, returns only leaf column items.
     *   When omitted, returns all top-level items regardless of type.
     * @param option.visible When `true`, returns only visible items. When `false`, returns only hidden items.
     *   Note that group items have no visibility toggle and are excluded when this filter is active.
     */
    findOptions(option = {}) {
        /* istanbul ignore next: :has() selector not supported in JSDOM */ if (option.group === true) {
            // Only group items — identified by the data-item-type="group" wrapper inside the list item
            return this.getList().findAll(`li:has([data-item-type="group"])`).map(wrapper => new ContentDisplayOptionWrapper(wrapper.getElement()));
        }
        /* istanbul ignore next: :has() selector not supported in JSDOM */
        if (option.group === false) {
            // Only leaf column items — identified by the data-item-type="column" wrapper
            return this.getList().findAll(`li:has([data-item-type="column"])`).map(wrapper => new ContentDisplayOptionWrapper(wrapper.getElement()));
        }
        // No group filter — return all top-level items
        return this.getList().findItems().map(wrapper => new ContentDisplayOptionWrapper(wrapper.getElement()));
    }
    /**
     * Returns the text filter input.
     */
    findTextFilter() {
        return this.findComponent(`.${styles_selectors_js_1.default['content-display-text-filter']}`, text_filter_1.default);
    }
    /**
     * Returns no match with the clear filter button.
     */
    findNoMatch() {
        return this.findByClassName(styles_selectors_js_1.default['content-display-no-match']);
    }
}
ContentDisplayPreferenceWrapper.rootSelector = styles_selectors_js_1.default['content-display'];
exports.default = ContentDisplayPreferenceWrapper;
