// @cloudscape-design/components/test-utils/selectors/prompt-input/index.js
"use strict";
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptInputMenuWrapper = void 0;
const selectors_1 = require("@cloudscape-design/test-utils-core/selectors");
const utils_1 = require("@cloudscape-design/test-utils-core/utils");
const option_1 = require("../internal/option");
const styles_selectors_js_1 = require("../../../dropdown/styles.selectors.js");
const styles_selectors_js_2 = require("../../../internal/components/selectable-item/styles.selectors.js");
const styles_selectors_js_3 = require("../../../prompt-input/test-classes/styles.selectors.js");
class PromptInputMenuWrapper extends selectors_1.ComponentWrapper {
    findOptions() {
        return this.findAll(`.${styles_selectors_js_2.default['selectable-item']}[data-test-index] .${option_1.default.rootSelector}`).map((elementWrapper) => new option_1.default(elementWrapper.getElement()));
    }
    /**
     * Returns an option from the menu.
     *
     * @param optionIndex 1-based index of the option to select.
     */
    findOption(optionIndex) {
        return this.findComponent(`.${styles_selectors_js_2.default['selectable-item']}[data-test-index="${optionIndex}"] .${option_1.default.rootSelector}`, option_1.default);
    }
    /**
     * Returns an option from the menu by its value
     *
     * @param value The 'value' of the option.
     */
    findOptionByValue(value) {
        const toReplace = (0, utils_1.escapeSelector)(value);
        return this.findComponent(`.${styles_selectors_js_2.default['selectable-item']} .${option_1.default.rootSelector}[data-value="${toReplace}"]`, option_1.default);
    }
}
exports.PromptInputMenuWrapper = PromptInputMenuWrapper;
class PromptInputWrapper extends selectors_1.ComponentWrapper {
    /**
     * Finds the native textarea element.
     *
     * Note: When `menus` or `tokens` is defined, the component uses a contentEditable element instead of a textarea.
     * In this case, use findContentEditableElement() instead.
     */
    findNativeTextarea() {
        return this.findByClassName(styles_selectors_js_3.default.textarea);
    }
    /**
     * Finds the contentEditable element used when `menus` or `tokens` is defined.
     */
    findContentEditableElement() {
        return this.findByClassName(styles_selectors_js_3.default['content-editable']);
    }
    /**
     * Finds the action button. Note that, despite its typings, this may return null.
     */
    findActionButton() {
        return this.findByClassName(styles_selectors_js_3.default['action-button']);
    }
    /**
     * Finds the secondary actions slot. Note that, despite its typings, this may return null.
     */
    findSecondaryActions() {
        return this.findByClassName(styles_selectors_js_3.default['secondary-actions']);
    }
    findSecondaryContent() {
        return this.findByClassName(styles_selectors_js_3.default['secondary-content']);
    }
    findCustomPrimaryAction() {
        return this.findByClassName(styles_selectors_js_3.default['primary-action']);
    }
    /**
     * Finds the menu dropdown when `menus` or `tokens` is defined.
     */
    findOpenMenu() {
        return (0, selectors_1.createWrapper)().findComponent(`.${styles_selectors_js_1.default.dropdown}[data-open=true]`, PromptInputMenuWrapper);
    }
}
PromptInputWrapper.rootSelector = styles_selectors_js_3.default.root;
exports.default = PromptInputWrapper;
