// @cloudscape-design/components/test-utils/dom/prompt-input/index.js
"use strict";
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptInputMenuWrapper = void 0;
const dom_1 = require("@cloudscape-design/test-utils-core/dom");
const utils_1 = require("@cloudscape-design/test-utils-core/utils");
const utils_dom_1 = require("@cloudscape-design/test-utils-core/utils-dom");
const option_1 = require("../internal/option");
const styles_selectors_js_1 = require("../../../dropdown/styles.selectors.js");
const styles_selectors_js_2 = require("../../../internal/components/selectable-item/styles.selectors.js");
const styles_selectors_js_3 = require("../../../prompt-input/test-classes/styles.selectors.js");
class PromptInputMenuWrapper extends dom_1.ComponentWrapper {
    findOptions() {
        return this.findAll(`.${styles_selectors_js_2.default['selectable-item']}[data-test-index] .${option_1.default.rootSelector}`).map((elementWrapper) => new option_1.default(elementWrapper.getElement()));
    }
    /**
     * Returns an option from the menu.
     *
     * @param optionIndex 1-based index of the option to select.
     */
    findOption(optionIndex) {
        return this.findComponent(`.${styles_selectors_js_2.default['selectable-item']}[data-test-index="${optionIndex}"] .${option_1.default.rootSelector}`, option_1.default);
    }
    /**
     * Returns an option from the menu by its value
     *
     * @param value The 'value' of the option.
     */
    findOptionByValue(value) {
        const toReplace = (0, utils_1.escapeSelector)(value);
        return this.findComponent(`.${styles_selectors_js_2.default['selectable-item']} .${option_1.default.rootSelector}[data-value="${toReplace}"]`, option_1.default);
    }
}
exports.PromptInputMenuWrapper = PromptInputMenuWrapper;
class PromptInputWrapper extends dom_1.ComponentWrapper {
    /**
     * Finds the native textarea element.
     *
     * Note: When `menus` or `tokens` is defined, the component uses a contentEditable element instead of a textarea.
     * In this case, use findContentEditableElement() instead.
     */
    findNativeTextarea() {
        return this.findByClassName(styles_selectors_js_3.default.textarea);
    }
    /**
     * Finds the contentEditable element used when `menus` or `tokens` is defined.
     */
    findContentEditableElement() {
        return this.findByClassName(styles_selectors_js_3.default['content-editable']);
    }
    /**
     * Finds the action button. Note that, despite its typings, this may return null.
     */
    findActionButton() {
        return this.findByClassName(styles_selectors_js_3.default['action-button']);
    }
    /**
     * Finds the secondary actions slot. Note that, despite its typings, this may return null.
     */
    findSecondaryActions() {
        return this.findByClassName(styles_selectors_js_3.default['secondary-actions']);
    }
    findSecondaryContent() {
        return this.findByClassName(styles_selectors_js_3.default['secondary-content']);
    }
    findCustomPrimaryAction() {
        return this.findByClassName(styles_selectors_js_3.default['primary-action']);
    }
    /**
     * Finds the menu dropdown when `menus` or `tokens` is defined.
     */
    findOpenMenu() {
        return (0, dom_1.createWrapper)().findComponent(`.${styles_selectors_js_1.default.dropdown}[data-open=true]`, PromptInputMenuWrapper);
    }
    /**
     * Gets the value of the component.
     *
     * Returns the current value of the textarea.
     *
     * When `menus` or `tokens` is defined, the component uses a contentEditable element.
     * Use findContentEditableElement().getElement().textContent instead.
     */
    getTextareaValue() {
        const textarea = this.findNativeTextarea();
        return textarea ? textarea.getElement().value : '';
    }
    /**
     * Sets the value of the textarea and calls the onChange handler.
     *
     * When `menus` or `tokens` is defined, the component uses a contentEditable element
     * and this method will have no effect.
     *
     * @param value value to set the textarea to.
     */
    setTextareaValue(value) {
        const textarea = this.findNativeTextarea();
        if (textarea) {
            const element = textarea.getElement();
            (0, utils_dom_1.act)(() => {
                const event = new Event('change', { bubbles: true, cancelable: false });
                (0, utils_dom_1.setNativeValue)(element, value);
                element.dispatchEvent(event);
            });
        }
    }
    /**
     * Checks if the menu is currently open.
     */
    isMenuOpen() {
        const menu = this.findOpenMenu();
        return menu !== null;
    }
    /**
     * Selects an option from the menu by simulating mouse events.
     *
     * @param value value of option to select
     */
    selectMenuOptionByValue(value) {
        (0, utils_dom_1.act)(() => {
            const menu = this.findOpenMenu();
            if (!menu) {
                throw new Error('Menu not found');
            }
            const option = menu.findOptionByValue(value);
            if (!option) {
                throw new Error(`Option with value "${value}" not found in menu`);
            }
            option.fireEvent(new MouseEvent('mouseup', { bubbles: true }));
        });
    }
    /**
     * Selects an option from the menu by simulating mouse events.
     *
     * @param optionIndex 1-based index of the option to select
     */
    selectMenuOption(optionIndex) {
        (0, utils_dom_1.act)(() => {
            const menu = this.findOpenMenu();
            if (!menu) {
                throw new Error('Menu not found');
            }
            const option = menu.findOption(optionIndex);
            if (!option) {
                throw new Error(`Option at index ${optionIndex} not found in menu`);
            }
            option.fireEvent(new MouseEvent('mouseup', { bubbles: true }));
        });
    }
}
PromptInputWrapper.rootSelector = styles_selectors_js_3.default.root;
exports.default = PromptInputWrapper;
__decorate([
    dom_1.usesDom
], PromptInputWrapper.prototype, "getTextareaValue", null);
__decorate([
    dom_1.usesDom
], PromptInputWrapper.prototype, "setTextareaValue", null);
__decorate([
    dom_1.usesDom
], PromptInputWrapper.prototype, "isMenuOpen", null);
__decorate([
    dom_1.usesDom
], PromptInputWrapper.prototype, "selectMenuOptionByValue", null);
__decorate([
    dom_1.usesDom
], PromptInputWrapper.prototype, "selectMenuOption", null);
